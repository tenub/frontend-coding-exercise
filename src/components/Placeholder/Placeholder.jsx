import React from 'react';
import { DropTarget } from 'react-dnd';

import { ItemTypes } from '../../ItemTypes';

class Placeholder extends React.PureComponent {
  /**
   * Only process item types defined here whenever an item is dropped
   * onto the Placeholder.
   */
  static types = [
    ItemTypes.BUTTON,
    ItemTypes.INPUT,
    ItemTypes.TEXT,
  ]

  static spec = {
    drop: (props, monitor) => {
      const { index } = props;
      return {
        name: 'Placeholder',
        index,
      };
    },
  }

  static collect = (connect, monitor) => ({
    canDrop: monitor.canDrop(),
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
  })

  render() {
    const { children, isOver, canDrop, connectDropTarget } = this.props;
    const isActive = canDrop && isOver;

    let backgroundColor;
    if (isActive) {
      backgroundColor = 'grey';
    } else if (canDrop) {
      backgroundColor = 'lightgrey';
    } else {
      backgroundColor = 'white'
    }

    return connectDropTarget(
      <div
        className='element-placeholder'
        style={{ backgroundColor }}
      >
        {children}
      </div>
    );
  }
}

export default DropTarget(
  Placeholder.types,
  Placeholder.spec,
  Placeholder.collect,
)(Placeholder);
