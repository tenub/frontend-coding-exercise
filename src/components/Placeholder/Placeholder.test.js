import React from 'react';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { DndProvider } from 'react-dnd';
import TestRenderer from 'react-test-renderer';

import Placeholder from './';

test('Placeholder renders successfully', () => {
  const placeholder = (
    <DndProvider backend={HTML5Backend}>
      <Placeholder index={0}>
        Text
      </Placeholder>
    </DndProvider>
  );
  const testRenderer = TestRenderer.create(placeholder);
  const testTree = testRenderer.toJSON();
  expect(testTree).toMatchSnapshot();
});
