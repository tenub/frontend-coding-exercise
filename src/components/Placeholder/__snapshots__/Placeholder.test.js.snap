// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Placeholder renders successfully 1`] = `
<div
  className="element-placeholder"
  style={
    Object {
      "backgroundColor": "white",
    }
  }
>
  Text
</div>
`;
